files/private-chef-ctl-commands/key-control.rb

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2014 Chef Software, Inc.
+# All Rights Reserved
+
+require "sequel"
+
+cmd_args = ARGV[3..-1]
+
+# Pull all the DB params from config; the database name seems to be hardcoded,
+# however -- I found it in the postgresql.rb recipe in the private-chef cookbooks;
+# I'm not sure why we hardcode it instead of making it configurable, but I assume
+# there are reasons?
+db_user = running_config['private_chef']['postgresql']['sql_user']
+db_password = running_config['private_chef']['postgresql']['sql_password']
+db_host = running_config['private_chef']['postgresql']['listen_address']
+
+# set convert_infinite_timestamps to a float so that Sequel doesn't crash out on querying an infinity timestamp
+# see this commit for more details
+# https://github.com/jeremyevans/sequel/commit/2440e98b4a7063a90d8337e57b1c3e3d3a1cbd0b
+DB = Sequel.connect(:adapter => 'postgres', :host => db_host,
+  :database => 'opscode_chef', :user => db_user, :password => db_password, :convert_infinite_timestamps => :float)
+
+# Helper Class
+class PublicKeyOutputter
+  def output_key_results(results, hide_public_keys)
+    results.each do |row|
+      puts "key_id: #{row[:key_id]}"
+      unless hide_public_keys
+        puts "    public_key:"
+        row[:public_key].each_line do |line|
+          puts "    #{line}"
+        end
+      end
+    end
+  end
+end
+
+add_command_under_category "key-create", "key-rotation", "Create a new key", 2 do
+  # Okay, obviously, this doesn't do what it says, this is just a random simple
+  # query to verify that the database connection is functional; on dev-vm it
+  # will produce some small number like "1" depending on what you've been doing
+  DB.fetch("SELECT COUNT(*) FROM USER") do |row|
+    puts row[:count]
+  end
+end
+
+# add_command_under_category "key-delete", "key-rotation", "Delete a key", 2 do
+#   TODO: some stuff
+# end
+
+add_command_under_category "list-client-keys", "key-rotation", "List keys for a user.", 2 do
+  # allow users to hide public key from output, this will allow for much easier
+  # viewing of keys if a user has lots of keys and they just want to see the key_ids
+  hide_public_keys = cmd_args[2] == "--hide-public-keys"
+  unless cmd_args.length == 2 || (cmd_args.length == 3 && cmd_args[2] == "--hide-public-keys")
```

Comment:
Actually, for the upgrade commands for Chef 12 we set one up. That would be available here with the correct imports. It's easiest to look at the upgrade code.

(This wouldn't be in base omnibus-ctl, however, but all the pieces are in opscode-omnibus)